test/py/test_device_set_irqs.py

```diff
@@ -0,0 +1,345 @@
+    sock.send(hdr + payload)
+    vfu_run_ctx(ctx)
+    get_reply(sock, expect=errno.EINVAL)
+
+def test_device_set_irqs_bad_start_for_count_0():
+    # struct vfio_irq_set
+    payload = struct.pack("IIIII", 20, VFIO_IRQ_SET_ACTION_MASK |
+                          VFIO_IRQ_SET_DATA_NONE, VFU_DEV_MSIX_IRQ, 1, 0)
+
+    hdr = vfio_user_header(VFIO_USER_DEVICE_SET_IRQS, size=len(payload))
+    sock.send(hdr + payload)
+    vfu_run_ctx(ctx)
+    get_reply(sock, expect=errno.EINVAL)
+
+def test_device_set_irqs_bad_action_for_count_0():
+    # struct vfio_irq_set
+    payload = struct.pack("IIIII", 20, VFIO_IRQ_SET_ACTION_MASK |
+                          VFIO_IRQ_SET_DATA_NONE, VFU_DEV_MSIX_IRQ, 0, 0)
+
+    hdr = vfio_user_header(VFIO_USER_DEVICE_SET_IRQS, size=len(payload))
+    sock.send(hdr + payload)
+    vfu_run_ctx(ctx)
+    get_reply(sock, expect=errno.EINVAL)
+
+def test_device_set_irqs_bad_action_and_data_type_for_count_0():
+    # struct vfio_irq_set
+    payload = struct.pack("IIIII", 20, VFIO_IRQ_SET_ACTION_TRIGGER |
+                          VFIO_IRQ_SET_DATA_BOOL, VFU_DEV_MSIX_IRQ, 0, 0)
+
+    hdr = vfio_user_header(VFIO_USER_DEVICE_SET_IRQS, size=len(payload))
+    sock.send(hdr + payload)
+    vfu_run_ctx(ctx)
+    get_reply(sock, expect=errno.EINVAL)
+
+def test_device_set_irqs_bad_fds_for_DATA_BOOL():
+    # struct vfio_irq_set
+    payload = struct.pack("IIIII", 20, VFIO_IRQ_SET_ACTION_TRIGGER |
+                          VFIO_IRQ_SET_DATA_BOOL, VFU_DEV_MSIX_IRQ, 0, 1)
+
+    fd = eventfd()
+
+    hdr = vfio_user_header(VFIO_USER_DEVICE_SET_IRQS, size=len(payload))
+    sock.sendmsg([hdr + payload], [(socket.SOL_SOCKET, socket.SCM_RIGHTS,
+                 struct.pack("I", fd))])
+    vfu_run_ctx(ctx)
+    get_reply(sock, expect=errno.EINVAL)
+
+    os.close(fd)
+
+def test_device_set_irqs_bad_fds_for_DATA_NONE():
+    # struct vfio_irq_set
+    payload = struct.pack("IIIII", 20, VFIO_IRQ_SET_ACTION_TRIGGER |
+                          VFIO_IRQ_SET_DATA_NONE, VFU_DEV_MSIX_IRQ, 0, 1)
+
+    fd = eventfd()
+
+    hdr = vfio_user_header(VFIO_USER_DEVICE_SET_IRQS, size=len(payload))
+    sock.sendmsg([hdr + payload], [(socket.SOL_SOCKET, socket.SCM_RIGHTS,
+                 struct.pack("I", fd))])
+    vfu_run_ctx(ctx)
+    get_reply(sock, expect=errno.EINVAL)
+
+    os.close(fd)
+
+def test_device_set_irqs_bad_fds_for_count_2():
+    # struct vfio_irq_set
+    payload = struct.pack("IIIII", 20, VFIO_IRQ_SET_ACTION_TRIGGER |
+                          VFIO_IRQ_SET_DATA_EVENTFD, VFU_DEV_MSIX_IRQ, 0, 2)
+
+    fd = eventfd()
+
+    hdr = vfio_user_header(VFIO_USER_DEVICE_SET_IRQS, size=len(payload))
+    sock.sendmsg([hdr + payload], [(socket.SOL_SOCKET, socket.SCM_RIGHTS,
+                 struct.pack("I", fd))])
+    vfu_run_ctx(ctx)
+    get_reply(sock, expect=errno.EINVAL)
+
+    os.close(fd)
+
+def test_device_set_irqs_disable():
+    # struct vfio_irq_set
```

Comment:
Can we have a very short explanation what is achieved with each command? I recall the IRQ API is not trivial.

Reply:
I'm not sure I want to repeat what the spec says? That should hopefully be sufficient?

Reply:
I was thinking something super short like "start=0 count=0 means disable all IRQs" but if you think it's redundant we can skip it